File: src/test/playwright/pages/BasePage.js
/**
 * Base Page class with common functionality for all pages
 */
class BasePage {
  constructor(page) {
    this.page = page;
  }

  /**
   * Navigate to a specific URL
   */
  async goto(url) {
    await this.page.goto(url);
  }

  /**
   * Get page title
   */
  async getTitle() {
    return await this.page.title();
  }

  /**
   * Check if navigation menu is visible
   */
  async isNavigationMenuVisible() {
    return await this.page.locator('nav.navbar').isVisible();
  }

  /**
   * Click on Home link in navigation
   */
  async clickHome() {
    await this.page.locator('nav.navbar a[href="/"]').click();
  }

  /**
   * Click on Find Owners link in navigation
   */
  async clickFindOwners() {
    await this.page.locator('nav.navbar a[href="/owners/find"]').click();
  }

  /**
   * Click on Veterinarians link in navigation
   */
  async clickVeterinarians() {
    await this.page.locator('nav.navbar a[href="/vets.html"]').click();
  }

  /**
   * Click on Error link in navigation
   */
  async clickError() {
    await this.page.locator('nav.navbar a[href="/oups"]').click();
  }

  /**
   * Wait for page to load completely
   */
  async waitForLoad() {
    await this.page.waitForLoadState('networkidle');
  }

  /**
   * Check if success message is visible
   */
  async isSuccessMessageVisible() {
    return await this.page.locator('.alert-success').isVisible();
  }

  /**
   * Check if error message is visible
   */
  async isErrorMessageVisible() {
    return await this.page.locator('.alert-danger').isVisible();
  }

  /**
   * Get success message text
   */
  async getSuccessMessage() {
    return await this.page.locator('.alert-success').textContent();
  }

  /**
   * Get error message text
   */
  async getErrorMessage() {
    return await this.page.locator('.alert-danger').textContent();
  }
}

module.exports = { BasePage };

File: src/test/java/org/springframework/samples/petclinic/playwright/pages/BasePage.java
/*
 * Copyright 2012-2025 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.springframework.samples.petclinic.playwright.pages;

import com.microsoft.playwright.Page;
import com.microsoft.playwright.options.LoadState;

/**
 * Base page class for Page Object Model pattern implementation. Contains common
 * functionality shared across all pages.
 */
public abstract class BasePage {

	protected final Page page;

	public BasePage(Page page) {
		this.page = page;
	}

	/**
	 * Wait for the page to fully load
	 */
	public void waitForPageLoad() {
		page.waitForLoadState(LoadState.NETWORKIDLE);
	}

	/**
	 * Get the page title
	 */
	public String getTitle() {
		return page.title();
	}

	/**
	 * Get the current URL
	 */
	public String getCurrentUrl() {
		return page.url();
	}

	/**
	 * Check if an element is visible using a specific locator
	 */
	public boolean isElementVisible(String locator) {
		return page.isVisible(locator);
	}

	/**
	 * Click on navigation menu item by text
	 */
	public void clickNavigationItem(String itemText) {
		// Using specific navigation structure - nav.navbar contains menu items
		page.locator("nav.navbar").locator("text=" + itemText).click();
		waitForPageLoad();
	}

	/**
	 * Verify navigation bar is present
	 */
	public boolean isNavigationBarVisible() {
		return page.locator("nav.navbar").isVisible();
	}

	/**
	 * Get the main navigation menu locator for reliable element identification
	 */
	protected String getNavigationSelector() {
		return "nav.navbar";
	}

}